// ra-ui-materialui/dist/esm/theme/nanoTheme.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { createTheme } from '@mui/material';
/**
 * Nano: A dense theme with minimal chrome, ideal for complex apps.
 *
 * Uses a small font size, reduced spacing, text buttons, standard variant inputs, pale colors.
 */
var componentsOverrides = function (theme) {
    var _a;
    return ({
        MuiAlert: {
            defaultProps: {
                variant: 'outlined',
            },
        },
        MuiAppBar: {
            defaultProps: {
                elevation: 1,
            },
        },
        MuiAutocomplete: {
            variants: [
                {
                    props: {},
                    style: function (_a) {
                        var _b;
                        var theme = _a.theme;
                        return (_b = {},
                            _b[theme.breakpoints.down('sm')] = { width: '100%' },
                            _b);
                    },
                },
            ],
            styleOverrides: {
                root: {
                    '& label+.MuiInput-root.MuiInputBase-root': {
                        marginTop: theme.spacing(1.5),
                    },
                    '& label[data-shrink=false]+.MuiInput-root.MuiInputBase-root': {
                        marginTop: 0,
                        paddingBottom: theme.spacing(2),
                    },
                },
                input: {
                    padding: theme.spacing(0.5),
                },
            },
        },
        MuiButton: {
            defaultProps: {
                variant: 'text',
                size: 'small',
            },
            styleOverrides: {
                root: {
                    paddingTop: theme.spacing(0.2),
                    paddingBottom: theme.spacing(0.2),
                },
            },
        },
        MuiCard: {
            defaultProps: {
                square: true,
            },
        },
        MuiChip: {
            defaultProps: {
                variant: 'outlined',
            },
        },
        MuiFormControl: {
            defaultProps: {
                variant: 'standard',
                margin: 'dense',
                size: 'small',
            },
        },
        MuiFormHelperText: {
            defaultProps: {
                margin: 'dense',
            },
        },
        MuiIconButton: {
            defaultProps: {
                size: 'small',
            },
        },
        MuiInputBase: {
            styleOverrides: {
                root: {
                    'label+&.MuiInputBase-root': {
                        marginTop: theme.spacing(1.5),
                    },
                    'label[data-shrink=false]+&.MuiInputBase-root': {
                        marginTop: 0,
                        paddingBottom: theme.spacing(1.5),
                    },
                },
                input: {
                    padding: theme.spacing(0.5),
                },
            },
        },
        MuiInputLabel: {
            styleOverrides: {
                root: {
                    paddingLeft: theme.spacing(0.5),
                },
            },
            defaultProps: {
                margin: 'dense',
            },
        },
        MuiListItem: {
            defaultProps: {
                dense: true,
            },
        },
        MuiListItemIcon: {
            styleOverrides: {
                root: {
                    '&.MuiListItemIcon-root': {
                        minWidth: theme.spacing(3.5),
                    },
                },
            },
        },
        MuiMenuItem: {
            styleOverrides: {
                root: {
                    paddingTop: theme.spacing(0.5),
                    paddingBottom: theme.spacing(0.5),
                    paddingLeft: theme.spacing(1),
                    paddingRight: theme.spacing(1),
                },
            },
        },
        MuiOutlinedInput: {
            defaultProps: {
                margin: 'dense',
            },
            styleOverrides: {
                input: {
                    padding: 16,
                },
            },
        },
        MuiPaper: {
            styleOverrides: {
                elevation1: {
                    boxShadow: theme.shadows[1],
                },
                root: {
                    backgroundColor: theme.palette.background.default,
                },
            },
        },
        MuiSnackbar: {
            styleOverrides: {
                root: {
                    '& .RaNotification-error': {
                        border: "1px solid ".concat(theme.palette.error.main),
                        backgroundColor: "".concat(theme.palette.common.white, " !important"),
                        color: "".concat(theme.palette.error.main, " !important"),
                    },
                    '& .RaNotification-warning': {
                        border: "1px solid ".concat(theme.palette.warning.main),
                        backgroundColor: "".concat(theme.palette.common.white, " !important"),
                        color: "".concat(theme.palette.warning.main, " !important"),
                    },
                    '& .RaNotification-info': {
                        border: "1px solid ".concat(theme.palette.info.main),
                        backgroundColor: "".concat(theme.palette.common.white, " !important"),
                        color: "".concat(theme.palette.info.main, " !important"),
                    },
                    '& .RaNotification-success': {
                        border: "1px solid ".concat(theme.palette.success.main),
                        backgroundColor: "".concat(theme.palette.common.white, " !important"),
                        color: "".concat(theme.palette.success.main, " !important"),
                    },
                },
            },
        },
        MuiTabs: {
            styleOverrides: {
                root: {
                    '&.MuiTabs-root': {
                        minHeight: theme.spacing(3.5),
                    },
                },
            },
        },
        MuiTab: {
            styleOverrides: {
                root: {
                    '&.MuiTab-root': {
                        padding: "".concat(theme.spacing(0.5), " ").concat(theme.spacing(1)),
                        minHeight: theme.spacing(3.5),
                        minWidth: theme.spacing(10),
                    },
                },
            },
        },
        MuiTable: {
            defaultProps: {
                size: 'small',
            },
        },
        MuiTableCell: {
            styleOverrides: {
                root: {
                    padding: theme.spacing(1),
                    '&.MuiTableCell-sizeSmall': {
                        padding: theme.spacing(0.5),
                    },
                    '&.MuiTableCell-paddingNone': {
                        padding: 0,
                    },
                },
            },
        },
        MuiTextField: {
            defaultProps: {
                variant: 'standard',
                margin: 'dense',
                size: 'small',
            },
            variants: [
                {
                    props: {},
                    style: function (_a) {
                        var _b;
                        var theme = _a.theme;
                        return (_b = {},
                            _b[theme.breakpoints.down('sm')] = { width: '100%' },
                            _b);
                    },
                },
            ],
        },
        MuiToolbar: {
            defaultProps: {
                variant: 'dense',
            },
            styleOverrides: {
                root: {
                    minHeight: theme.spacing(4.5),
                },
                regular: {
                    backgroundColor: theme.palette.background.paper,
                },
            },
        },
        RaDatagrid: {
            styleOverrides: {
                root: {
                    '& .RaDatagrid-headerCell': {
                        color: theme.palette.primary.main,
                    },
                },
            },
        },
        RaFilterForm: {
            styleOverrides: {
                root: (_a = {},
                    _a[theme.breakpoints.up('sm')] = {
                        minHeight: theme.spacing(7.1),
                    },
                    _a),
            },
        },
        RaFilterFormInput: {
            styleOverrides: {
                root: {
                    '& .RaFilterFormInput-hideButton': {
                        marginBottom: theme.spacing(0.5),
                    },
                },
            },
        },
        RaLayout: {
            styleOverrides: {
                root: {
                    '& .RaLayout-appFrame': {
                        marginTop: theme.spacing(5),
                    },
                },
            },
        },
        RaLoadingIndicator: {
            styleOverrides: {
                root: {
                    '& .RaLoadingIndicator-loader': {
                        top: '20%',
                        left: '20%',
                    },
                },
            },
        },
        RaMenuItemLink: {
            styleOverrides: {
                root: {
                    paddingLeft: theme.spacing(1),
                    paddingRight: theme.spacing(1),
                    '&.RaMenuItemLink-active': {
                        color: theme.palette.primary.dark,
                        fontWeight: 700,
                        '& .MuiSvgIcon-root': {
                            fill: theme.palette.primary.dark,
                        },
                    },
                },
            },
        },
    });
};
var alert = {
    error: { main: '#B57185' },
    warning: { main: '#F2CB05' },
    info: { main: '#39AEA9' },
    success: { main: '#00745F' },
};
var darkPalette = __assign({ mode: 'dark', primary: { main: '#f9fafb' }, secondary: { main: '#a0a0a0' }, background: { default: '#363D40' } }, alert);
var lightPalette = __assign({ mode: 'light', primary: { main: '#00585C' }, secondary: { main: '#64B4B8' }, background: { default: '#f9fafb' }, text: { primary: '#212b36' } }, alert);
var createNanoTheme = function (palette) {
    var themeOptions = {
        palette: palette,
        shape: { borderRadius: 0 },
        sidebar: {
            width: 200,
            closedWidth: 36,
        },
        spacing: 8,
        typography: {
            fontFamily: 'Onest, sans-serif',
            fontSize: 12,
            h1: { fontSize: '7rem' },
            h2: { fontWeight: 400 },
            h3: { fontWeight: 500 },
            h4: { fontWeight: 700 },
            h5: { fontWeight: 700 },
        },
    };
    var theme = createTheme(themeOptions);
    theme.components = componentsOverrides(theme);
    return theme;
};
export var nanoLightTheme = createNanoTheme(lightPalette);
export var nanoDarkTheme = createNanoTheme(darkPalette);
